gms/feature.hh

```diff
@@ -145,6 +145,7 @@ extern const std::string_view PER_TABLE_CACHING;
 extern const std::string_view DIGEST_FOR_NULL_VALUES;
 extern const std::string_view CORRECT_IDX_TOKEN_IN_SECONDARY_INDEX;
 extern const std::string_view ALTERNATOR_STREAMS;
+extern const std::string_view QUERY_STATUS_FOR_READS;
```

Comment:
Commenting here because I can't comment on the entire patch :-(
I didn't 'understand why a flag is needed. When we add a new optional part at the end of RPC requests or replies, we already have a mechanism to support a mixture of old and new nodes - an old node will ignore the "status" field sent by a new node, and a new node will know it didn't get a status from an old node.  So why do we need the extra cluster-wide flag? It would be nice to explain this in the commit message.

Reply:
https://github.com/scylladb/scylla/pull/7773#discussion_r539194347

Reply:
It looks like you have two different mechanisms for indicating failure - the pre-existing error, and the new "status". Why do we need both mechanisms?
Also, even if you do have both, if an error would produce *both* an error and a status, even an old node would understand that this is a failure. No?

Reply:
There's no way of passing an error type via Seastar rpc, and following the discussion here https://github.com/scylladb/seastar/issues/835 , I went with propagating the error type as a field. If I get it correctly, the old error is generally supposed to be an error related to RPC, as in "calling the remote procedure failed", while the status field has application logic - e.g. that a read failed because a replica got overloaded.
I'm pretty sure that the RPC layer allows sending either data or an error, but not both. If I'm not correct, then your suggestion is applicable. @gleb-cloudius?

Reply:
I have no idea what @nyh has in mind. If he have a concrete proposal he need to expend on it. Obviously if you want to return something else instead of an error you need to make sure that the receiving side can interpret this something correctly.

Reply:
My thinking was that just like in Seastar both the success and the failure have types (i.e., the exception contains an object, not just the fact that an exception happened), so should RPC have a way to say which error happened, and I was. Without this feature, now when Piotr wants to return a special type of error, he is forced to return a success (!) plus an extra result field which says error, but as we noticed the result is a mess - we can't use the new mechanism if the entire cluster doesn't support it (or old nodes will think the reply is a success, while it's meant to be a type of error), and it is undefined what exactly happens if the result contains both the actual result and an error code.

I realize we can't overhaul RPC now, just for this feature. However, I raised to Piotr one idea that could allow the read verb to return multiple types of errors without changing RPC. RPC errors already have string messaes. We could (ab)use the string message in the same way that HTTP does this: begin the string with a numeric error code. This can easily be made backward-compatible, because today - I assume - the string the existing code returns does not begin with a numeric error code, so if new nodes start to return numeric error codes then 1. old nodes will still consider those responses as errors, and 2. new nodes receving a response from an old node will know this is an error.

Reply:
Piotr is not forced to return a success. He can introduce another verb that returns variant<result, error> but that requires introducing another verb. This and why it is not RPC job to transport user exception is discussed in https://github.com/scylladb/seastar/issues/835. Did you read it?

I cannot really consider using error string as a way to transport error codes seriously. It is fine for hacking a quick prototype, but not a proper protocol.

Reply:
Adding a new verb isn't a problem. It requires hassle (cluster feature to know if we can call it), but it's better than a hack that we'll have to work around forever afterwards.

Reply:
Which hack are you talking about?

Reply:
It's a "hack" to have a verb return both the success and error code, and hope nobody returns **both** of them because we don't know what that is supposed to mean.

Reply:
> I cannot really consider using error string as a way to transport error codes seriously. It is fine for hacking a quick prototype, but not a proper protocol.

Just out of curiosity, what did you imagine to be the use of these error strings, if we cannot use them to tell error types apart? If you don't like error codes as *part* of the string, how about returning different error strings (no parsing - just comparing the full string) - or is that also not a proper protocol.

Reply:
The verb does not return success. Where did you get that?  It is definitely not as clean as introducing another verb with variant<> return value, but it is much less hassle and I do not see why would we need to "work around it forever afterwards".

Reply:
> > I cannot really consider using error string as a way to transport error codes seriously. It is fine for hacking a quick prototype, but not a proper protocol.
> 
> Just out of curiosity, what did you imagine to be the use of these error strings, if we cannot use them to tell error types apart?
Logged for information purpose.

> If you don't like error codes as _part_ of the string, how about returning different error strings (no parsing - just comparing the full string) - or is that also not a proper protocol.

Of course not.